### PR TITLE
Wire GetSpeculativeTypeInfo with nullability information

### PR DESCRIPTION
**Pull request by 333fred**
@dotnet/roslyn-compiler @chsienki for review. /cc @jasonmalinowski @ryzngard 

Fixes https://github.com/dotnet/roslyn/issues/35037

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs [hunk 1]
```diff
@@ -135,6 +139,9 @@ internal override MemberSemanticModel GetMemberModel(SyntaxNode node)
             return IsInTree(node) ? this : null;
         }
 
+        // We hide this property, as EnsureRootBoundForNullabilityIfNecessary can cause
```
Comment (chsienki) [inline src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs]: Do we actually use this property in a lot of places? Properties with big side effects make me nervous. If its only used in one or two places, consider not calling `EnsureRootBoundForNullabilityIfNecessary` leaving it to the caller instead, and just assert that _lazySnapshotmanager is populated.

Comment (333fred) [reply src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs]: It's used in derived classes, so I'd prefer to keep the field private. I will change this to a method though, so the side effects aren't unexpected.

---
In reply to: [298375705](https://github.com/dotnet/roslyn/pull/36833#discussion_r298375705) [](ancestors = 298375705)

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler for a second review.

File: src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs [hunk 2]
```diff
@@ -27,6 +27,10 @@ internal abstract partial class MemberSemanticModel : CSharpSemanticModel
         private readonly Dictionary<SyntaxNode, ImmutableArray<BoundNode>> _guardedNodeMap = new Dictionary<SyntaxNode, ImmutableArray<BoundNode>>();
         private Dictionary<SyntaxNode, BoundStatement> _lazyGuardedSynthesizedStatementsMap;
         private NullableWalker.SnapshotManager _lazySnapshotManager;
+        /// <summary>
+        /// Only used when this is a speculative semantic model.
+        /// </summary>
+        private NullableWalker.SnapshotManager _parentSnapshotManagerOpt;
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs]: >private [](start = 8, length = 7)

`readonly`

File: src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs [hunk 3]
```diff
@@ -135,13 +139,13 @@ internal override MemberSemanticModel GetMemberModel(SyntaxNode node)
             return IsInTree(node) ? this : null;
         }
 
-        protected NullableWalker.SnapshotManager SnapshotManager
+        // We hide this property, as EnsureRootBoundForNullabilityIfNecessary can cause
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs]: >We hide this property, as [](start = 11, length = 25)

This is a little confusing since this member is not a property. Perhaps "Not a property because ...".

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs [hunk 4]
```diff
@@ -233,7 +245,7 @@ private Binder GetSpeculativeBinderForAttribute(int position, AttributeSyntax at
             return new ExecutableCodeBinder(attribute, binder.ContainingMemberOrLambda, binder).GetBinder(attribute);
         }
 
-        private static BoundExpression GetSpeculativelyBoundExpressionHelper(Binder binder, ExpressionSyntax expression, SpeculativeBindingOption bindingOption, DiagnosticBag diagnostics)
+        protected static BoundExpression GetSpeculativelyBoundExpressionHelper(Binder binder, ExpressionSyntax expression, SpeculativeBindingOption bindingOption, DiagnosticBag diagnostics)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs]: >protected [](start = 8, length = 9)

Is this change needed?

**Comment by cston** [comment]
>             private readonly ImmutableArray<SharedWalkerState> _walkerGlobalStates;

"Shared" here and throughout this file?

---
Refers to: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs:20 in dc39452. [](commit_id = dc39452826d1d807915e022bf70f3581bd8ed70a, deletion_comment = False)
